server/util/fastcopy/fastcopy_windows.go

```diff
@@ -2,16 +2,174 @@
+// On newer Windows versions, CopyFileW will automatically use block cloning
+// if the underlying filesystem supports it (e.g., ReFS, DevDrive).
+// Reference:
+//
+//	https://devblogs.microsoft.com/engineering-at-microsoft/copy-on-write-in-win32-api-early-access/
 func FastCopy(source, destination string) error {
-	err := os.Link(source, destination)
-	if !os.IsExist(err) {
+	lpExistingFileName, err := syscall.UTF16PtrFromString(source)
+	if err != nil {
+		return err
+	}
+	lpNewFileName, err := syscall.UTF16PtrFromString(destination)
+	if err != nil {
+		return err
+	}
+
+	var bFailIfExists uint32 = 1
+	r1, _, err := syscall.SyscallN(
+		procCopyFileW.Addr(),
+		uintptr(unsafe.Pointer(lpExistingFileName)),
+		uintptr(unsafe.Pointer(lpNewFileName)),
+		uintptr(bFailIfExists),
+	)
+	if r1 != 0 {
+		return nil
+	}
+	// Handle file exists case the same way as Linux and Darwin
+	if os.IsExist(err) {
+		return nil
+	}
+	return fmt.Errorf("failed CopyFileW Win32 call from '%s' to '%s': %s", source, destination, err)
+}
+
+// https://learn.microsoft.com/en-us/windows/win32/api/winioctl/ns-winioctl-duplicate_extents_data
+type duplicateExtentsData struct {
+	FileHandle       windows.Handle
+	SourceFileOffset int64
+	TargetFileOffset int64
+	ByteCount        int64
+}
+
+// Implementation is heavily inspired by Git LFS
+// References:
+//
+//   - https://github.com/microsoft/CopyOnWrite/
+//   - https://github.com/git-lfs/git-lfs/blob/main/tools/util_windows.go
+//   - https://github.com/nicokoch/reflink/blob/e8d93b465f5d9ad340cd052b64bbc77b8ee107e2/src/sys/windows.rs#L199
+func reflink(source, destination string) error {
+	src, err := os.Open(source)
+	if err != nil {
+		return err
+	}
+	defer src.Close()
+
+	dst, err := os.OpenFile(destination, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0o666)
+	if err != nil {
+		// Handle file exists case the same way as Linux and Darwin
+		if os.IsExist(err) {
+			return nil
+		}
+		return err
+	}
+	defer dst.Close()
+
+	srcStat, err := src.Stat()
+	if err != nil {
 		return err
 	}
-	return nil
+
+	fileSize := srcStat.Size()
+
+	err = dst.Truncate(fileSize) // set file size. There is a requirement "The destination region must not extend past the end of file."
+	if err != nil {
+		return err
+	}
+
+	offset := int64(0)
+
+	// Requirement
+	// * The source and destination regions must begin and end at a cluster boundary. (4KiB or 64KiB)
+	// * cloneRegionSize less than 4GiB.
+	// see https://docs.microsoft.com/windows/win32/fileio/block-cloning
+	// Clone first xGiB region.
+	for ; offset+GiB < fileSize; offset += GiB {
+		if err := callDuplicateExtentsToFile(dst, src, offset, GiB); err != nil {
+			return err
+		}
+	}
+
+	// Clone tail. First try with 64KiB round up, then fallback to 4KiB.
+	for _, cloneRegionSize := range ReFSClusterSize {
+		err = callDuplicateExtentsToFile(dst, src, offset, roundUp(fileSize-offset, cloneRegionSize))
+		if err == nil {
+			return nil
+		}
+	}
+
+	return err
```

Comment:
Since this `err` spans several lines it might be clearer to declare it before the loop, like `var tailErr error` and use that instead of `err`.